apis/crd/v1alpha1/cloudentityselector_webhook.go

```diff
@@ -218,28 +218,123 @@ func (r *CloudEntitySelector) validateMatchSections() error {
 	// In AWS, Vpc name(in vpcMatch section) with either vm id or vm name(in vmMatch section) is not supported
 	if cloudProviderType == AzureCloudProvider {
 		for _, m := range r.Spec.VMSelector {
-			if m.VpcMatch != nil {
-				if len(strings.TrimSpace(m.VpcMatch.MatchName)) != 0 {
-					return fmt.Errorf("matchName is not supported in vpcMatch, " +
-						"use matchID instead of matchName")
-				}
+			if m.VpcMatch != nil && len(strings.TrimSpace(m.VpcMatch.MatchName)) != 0 {
+				return fmt.Errorf("matchName is not supported in vpcMatch, " +
+					"use matchID instead of matchName")
 			}
 		}
 	} else {
 		for _, m := range r.Spec.VMSelector {
-			if m.VpcMatch != nil {
-				if len(strings.TrimSpace(m.VpcMatch.MatchName)) != 0 {
-					for _, vmMatch := range m.VMMatch {
-						if len(strings.TrimSpace(vmMatch.MatchID)) != 0 ||
-							len(strings.TrimSpace(vmMatch.MatchName)) != 0 {
-							return fmt.Errorf("vpc matchName with either vm matchID" +
-								" or vm matchName is not supported, use vpc matchID instead" +
-								" of vpc matchName")
-						}
+			if m.VpcMatch != nil && len(strings.TrimSpace(m.VpcMatch.MatchName)) != 0 {
+				for _, vmMatch := range m.VMMatch {
+					if len(strings.TrimSpace(vmMatch.MatchID)) != 0 ||
+						len(strings.TrimSpace(vmMatch.MatchName)) != 0 {
+						return fmt.Errorf("vpc matchName with either vm matchID" +
+							" or vm matchName is not supported, use vpc matchID instead" +
+							" of vpc matchName")
 					}
 				}
 			}
 		}
 	}
+
+	// Block ambiguous match combinations as it can result in unpredictable behavior wrt "agented" configuration.
+	// vpcIDOnlyMatch map holds VPC ID as key if the selector contains only vpcMatch matchID.
+	// vmIDOnlyMatch map holds VM ID as key if the selector contains only vmMatch matchID.
+	// vmNameOnlyMatch map holds VM Name as key if the selector contains only vmMatch matchName.
+	// vmIDWithVpcMatch map holds VM ID and VPC ID as key if the selector contains vpcMatch matchID and vmMatch matchID.
+	// vmNameWithVpcMatch map holds VM Name and VPC ID as key if the selector contains vpcMatch matchID and vmMatch matchName.
+	vpcIDOnlyMatch := make(map[string]struct{})
+	vmIDOnlyMatch := make(map[string]struct{})
+	vmNameOnlyMatch := make(map[string]struct{})
+	vmIDWithVpcMatch := make(map[string]struct{})
+	vmNameWithVpcMatch := make(map[string]struct{})
+
+	for _, m := range r.Spec.VMSelector {
```

Comment:
What would happen if only a VPC name is given in VmSelector, and agented flag is true?

Reply:
As I had pointed in an offline discussion, we can't have VPC name indexer for agented feature as we do not have VPC name annotation in VM CRD, adding VPC name in VM CRD annotation will need more changes, we need to get VPC name from VPC ID and then update VM CRD for VPC name annotation. We do not have VPC ID to VPC name mapping right(what we have is VPC name to VPC id mapping). Hence, i am blocking agented=true configuration at webhook level when VPC name is configured in a VmSelector.